### ISSUE TITLE
Introduce terms for programmingLanguage that represent name and version?

### ISSUE THREAD

**Issue by sdruskat** [codemeta/codemeta#171]
We had a discussion at the FORCE2017 hackathon whether (or more precisely *that*) `programmingLanguage` should be able to support *name* of a programming language and *version* of the programming language in order to support (machine-actionable) disambiguation.

This could avoid simultaneous (but incompatible) use of `python`, `Python2`, `Python 3.6.1`, etc.

schema.org's [`ComputerLanguage`](http://schema.org/ComputerLanguage) does not support this ATM, but it could be incorporated as new CodeMeta-specific terms, e.g., `programmingLanguageName` and `programmingLanguageVersion`.

**Comment by cboettig** [comment]
Right, I've had the same thought, but it's not obvious where a controlled vocabulary for language names would be.  To better frame the issue, it would be good to flush out the use cases.  For instance, we determined that machine-actionable *installation* of dependencies and language is probably beyond the scope of codemeta -- it might work in some contexts, but package management is messy enough in a single ecosystem.  A simpler use case would be to tabulate the frequency that different languages are being used, say, across software archived in zenodo, but this still leaves open the question of version etc.

**Comment by sdruskat** [comment]
For an earlier version of CFF I've used GitHub's language list at https://github.com/github/linguist/blob/master/lib/linguist/languages.yml which doesn't include versions as such, but lists interpreters (e.g., `python`, `python2`, `python3`) instead. That said, I think I've come across some languages for which versions are baked into the name...

**Comment by npch** [comment]
I think from a citation point of view, language version is not important when referencing software. In fact I might go as far as saying that programming language is not important at all in citation (I'm interested in the software, not what it's coded in). As an analogy, I don't care whether a paper has been written in Word or LaTeX, I only care about its contents and author (I suppose the one counterpoint is I might care if it was authored on a platform that was known to delete words randomly, but I think that's well out of scope for citation referencing). 

Obviously, programming language is important when describing software for the purpose of discovery as part of a subset e.g. give me all pieces of software with functionality X that are written in programming language Y. For this use case, the question for codemeta is the one that @moranegg raises: is version important for discovery or is language name enough? Or is there an in-between e.g. Python2 is distinct from Python3, FORTRAN is distinct from Fortran 90.

**Comment by sdruskat** [comment]
*Disclaimer: I might be overthinking this.*

I think you're right @npch, programming language is *not* important for citation use cases, at all.

For discovery use cases I think version *may* be important, e.g., for potential re-use (cf. Python 2 vs 3). For these cases, I'd prefer to have "name: Python, version: 2" vs just "Python2", simply because "Python2" is not the name of a programming language, it's "Python" and the version is "2". 

Baking versions into names may also cause a lot of clutter, so that you'd have "Python2", "python2.7", Python3.6.1", etc., so 3 "different" programming languages. Dividing name and version is a lot cleaner, and would also allow for, e.g., better parsing of versions (someone interested in a software written in *any* sub-version of Python 2 would be able to ignore anything after the first int of `version`).

I'm not sure whether a name/version split may cause downstream issues with crosswalking, although I expect that this is something that tooling would take care of?

**Comment by rebeccaringuette** [comment]
2nd the idea of keeping name and version separate. Could make the version an optional subfield.
